spec/function.dd

```diff
@@ -224,6 +222,26 @@ pure int foo(int i,
 }
 ---
 
+    $(P An implementation may assume that a `pure` function that (a) accepts
+    only parameters without mutable indirections, and (b) returns a result
+    without mutable indirections, will have the same effect for all invocation
+    with equivalent arguments, and is allowed to memoize the result of the
+    function. These are termed "strongly `pure`" functions in this document.
+    Note that a strongly pure function may still have behavior inconsistent with
+    memoization by e.g. using `cast`s or by changing behavior depending on the
+    address of its parameters. An implementation is currently not required to
+    enforce validity of memoization in all cases.)
+
+    $(P Any `pure` function that is not strongly pure cannot be memoized. The
+    compiler is required to honor all calls to the function, even if it appears
+    to do nothing. (Example: a pure function taking no arguments and returning
+    `int*` cannot be memoized. It also should not because it may be a typical
+    factory function returning a fresh pointer with each call.))
+
+    $(P `pure` functions returning `void` will be always called even if it is
+    strongly `pure`. The implementation must assume the function does something
+    outside the confines of the type system and is therefore not allowed to
+    elide the call, even if it appears to have no possible effect.)
```

Comment:
Wouldn't it be simpler to just say that `pure` functions returning `void` are not strongly `pure`?